src/hypothesis/extra/datetime.py

```diff
@@ -15,111 +15,412 @@
+               'otherwise be drawn.')
+        if not any(is_representable(min_datetime, tz) for tz in timezones):
+            raise InvalidArgument(msg % ('min_datetime', min_datetime))
+        if not any(is_representable(max_datetime, tz) for tz in timezones):
+            raise InvalidArgument(msg % ('max_datetime', max_datetime))
+
+    # If there is only a single possible date, `just` is more efficient
+    if allow_naive and not timezones:
+        if min_datetime == max_datetime:
+            return just(min_datetime)
+    elif not allow_naive and len(timezones) == 1:
+        to_tz = timezones[0]
+        return just(to_tz.normalize(min_datetime.astimezone(to_tz)))
+
     return DatetimeStrategy(
         allow_naive=allow_naive, timezones=timezones,
-        min_year=min_year, max_year=max_year,
+        min_datetime=min_datetime, max_datetime=max_datetime,
     )
 
 
 @defines_strategy
-def dates(min_year=None, max_year=None):
-    """Return a strategy for generating dates."""
-    return datetimes(
-        allow_naive=True, timezones=[],
-        min_year=min_year, max_year=max_year,
-    ).map(datetime_to_date)
+def dates(min_year=None, max_year=None, min_date=None, max_date=None):
+    """Return a strategy for generating dates.
 
+    The arguments are as for :py:func:`datetimes`, including the deprecation
+    of ``min_year`` and ``max_year``.  If ``min_date`` or ``max_date`` are
+    datetime objects, the time and any timezone information is discarded and
+    they are treated as dates.
 
-def datetime_to_date(dt):
-    return dt.date()
+    .. doctest::
+
+        >>> from hypothesis.extra.datetime import dates
+        >>> dates().example()  # doctest: +NORMALIZE_WHITESPACE
+        datetime.date(100, 12, 19)
+        >>> dates().example()  # doctest: +NORMALIZE_WHITESPACE
+        datetime.date(5621, 7, 9)
+
+    """
+    # Handle deprecation and conversion of min_year and max_year arguments
+    if (min_year is not None) or (max_year is not None):
+        if (min_date is not None) or (max_date is not None):
+            raise InvalidArgument(
+                'You cannot mix old-style (min_year/max_year) and new-style '
+                '(min_date/max_date) arguments to the date strategy.')
+        msg = ('The %s argument is deprecated.  Please use %s_date '
+               'with a datetime.date object instead.')
+        if min_year is not None:
+            note_deprecation(msg % ('min_year', 'min'))
+            min_date = dt.date.min.replace(year=min_year)
+        if max_year is not None:
+            note_deprecation(msg % ('max_year', 'max'))
+            max_date = dt.date.max.replace(year=max_year)
+    del min_year
+    del max_year
+    # This is part of the deprecation logic - we could just use these immutable
+    # objects as default arguments, but we need to distinguish whether a new
+    # argument was given by the user for deprecation errors.
```

Comment:
Eh, I think the convention in Python is to use `None` rather than a sentinel object.  It might be a little easier to make the API misbehave, but they're basically abusing the `Optional` type and the worst that can happen is they don't get the deprecation warning or validation error.